pkg/engine/overlay.go

```diff
@@ -0,0 +1,379 @@
+
+	kubepolicy "github.com/nirmata/kyverno/pkg/apis/policy/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ProcessOverlay handles validating admission request
+// Checks the target resourse for rules defined in the policy
+func ProcessOverlay(policy kubepolicy.Policy, rawResource []byte, gvk metav1.GroupVersionKind) ([]PatchBytes, []byte) {
+	var resource interface{}
+	json.Unmarshal(rawResource, &resource)
+
+	for _, rule := range policy.Spec.Rules {
+		if rule.Mutation == nil || rule.Mutation.Overlay == nil {
+			continue
+		}
+
+		ok := ResourceMeetsDescription(rawResource, rule.ResourceDescription, gvk)
+		if !ok {
+			log.Printf("Rule \"%s\" is not applicable to resource\n", rule.Name)
+			continue
+		}
+
+		overlay := *rule.Mutation.Overlay
+		if err, _ := applyOverlay(resource, overlay, ""); err != nil {
+			//return fmt.Errorf("%s: %s", *rule.Validation.Message, err.Error())
+		}
+	}
+
+	return nil, nil
+}
+
+func applyOverlay(resource, overlay interface{}, path string) ([]PatchBytes, error) {
+	var appliedPatches []PatchBytes
+
+	// resource item exists but has different type - replace
+	// all subtree within this path by overlay
+	if reflect.TypeOf(resource) != reflect.TypeOf(overlay) {
+		patch, err := replaceSubtree(overlay, path)
+		if err != nil {
+			return nil, err
+		}
+
+		appliedPatches = append(appliedPatches, patch)
+		return appliedPatches, nil
+	}
+
+	switch typedOverlay := overlay.(type) {
+	case map[string]interface{}:
+		typedResource := resource.(map[string]interface{})
+
+		for key, value := range typedOverlay {
+			if wrappedWithParentheses(key) {
+				key = key[1 : len(key)-1]
+			}
+			currentPath := path + key + "/"
+			resourcePart, ok := typedResource[key]
+
+			if ok {
+				patches, err := applyOverlay(resourcePart, value, currentPath)
+				if err != nil {
+					return nil, err
+				}
+
+				appliedPatches = append(appliedPatches, patches...)
+
+			} else {
+				patch, err := insertSubtree(value, currentPath)
+				if err != nil {
+					return nil, err
+				}
+
+				appliedPatches = append(appliedPatches, patch)
+			}
+		}
+	case []interface{}:
+		typedResource := resource.([]interface{})
+		patches, err := applyOverlayToArray(typedResource, typedOverlay, path)
+		if err != nil {
+			return nil, err
+		}
+
+		appliedPatches = append(appliedPatches, patches...)
+	case string:
+		patch, err := replaceSubtree(overlay, path)
+		if err != nil {
+			return nil, err
+		}
+
+		appliedPatches = append(appliedPatches, patch)
+	case float64:
+		patch, err := replaceSubtree(overlay, path)
+		if err != nil {
+			return nil, err
+		}
+
+		appliedPatches = append(appliedPatches, patch)
+	case int64:
+		patch, err := replaceSubtree(overlay, path)
+		if err != nil {
+			return nil, err
+		}
+
+		appliedPatches = append(appliedPatches, patch)
+	}
+
+	return appliedPatches, nil
+}
+
+func applyOverlayToArray(resource, overlay []interface{}, path string) ([]PatchBytes, error) {
```

Comment:
Add comments to the code
And the function is too big and contains many nested operation. Try to decompose it